### ISSUE TITLE
Compiling methods to webassembly

### ISSUE THREAD

**Issue by romankoho** [davideberly/GeometricTools#89]
Hello,

I just discovered this library during my search for minimal volume bounding box libraries and algorithms. The project I'm working on is written in **Javascript and ThreeJs**. I have a point cloud of 3-dimensional vectors and want to calculate the minimal volume bounding box. Would there be a way to run the MinimumVolumeBox3 (https://www.geometrictools.com/GTE/Mathematics/MinimumVolumeBox3.h) algorithm in the web?

I don't have any (professional) experience with C++ but I guess compiling it to WebAssembly might be the way to go. Did anyone do this with methods of this project already or could guide me a little on what would be the way to go?

thanks!

**Comment by davideberly** [comment]
I have no experience with WebAssembly. The Wikipedia page for the topic mentions various compilers to convert C++ to WebAssembly. However, I have no idea how one would include the output as part of a Javascript/ThreeJs project.

If you have a way of hooking up C# code to your project, the GeometricTools/Samples/CSharpCppManaged sample application has a managed layer to allow the C++ minimum-volume code to be exposed in a C# program.

**Comment by romankoho** [comment]
Hi David,
thanks for your quick reply. Depending on whether we find another solution I might try the WebAssembly way. If I find a solution I'll add to this discussion

**Comment by davideberly** [comment]
I downloaded Enscripten and managed to get a simple example working for minimum-volume bounding box. Unfortunately, when I set the number of threads to be positive in the query, the browser throws an exception. If I set the number of threads to zero, the browser displays the correct results. I do not know enough about web programming to understand why I cannot run the C++ code multithreaded.

I will post some code shortly with a description of what I did to get Enscripten to work, compile the code, and run it on a local server. I was unable to figure out how to specify a (local) file to load that has the input points. The std::ofstream call appeared to fail, even with a hard-coded path.

I wish web programming would be much simpler than what it is...

**Comment by davideberly** [comment]
@romankoho : I had an example working that loaded the points from a binary file. I tried it again later but now the file fails to open. I am not a web programmer, and the Emscripten tutorial regarding read/write files appears to be incomplete in that it fails the same way. (I am using a local server: python -m http.server)

That said, I uploaded to my website a zip file that has a simple cpp program with input points contained directly in the cpp file. All necessary GTE header files for compiling the minimum-volume box code are included in a subfolder in the zip file. There is a ReadMe.txt file that explains what I did (install Emscripten, get the environment set up, use emcc to compile). The output files are contained in the zip file as well as a screen capture that shows the minimum-volume box for a set of 16 points.

[MinimumVolumeBoxWASM.zip](https://www.geometrictools.com/github/MinimumVolumeBoxWASM.zip)
